service/frontend/http_api_server.go

```diff
@@ -0,0 +1,221 @@
+package frontend
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"go.temporal.io/api/proxy"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/rpc/encryption"
+	"go.temporal.io/server/common/util"
+)
+
+type HTTPAPIServer struct {
+	server            http.Server
+	listener          net.Listener
+	logger            log.Logger
+	serveMux          *runtime.ServeMux
+	shutdownDrainTime time.Duration
+	stopped           chan struct{}
+}
+
+func NewHTTPAPIServer(
+	serviceConfig *Config,
+	rpcConfig config.RPC,
+	grpcListener net.Listener,
+	rpcFactory common.RPCFactory,
+	tlsConfigProvider encryption.TLSConfigProvider,
+	logger log.Logger,
+) (*HTTPAPIServer, error) {
+	// Create a TCP listener the same as the frontend one but with different port
+	tcpAddrRef, _ := grpcListener.Addr().(*net.TCPAddr)
+	if tcpAddrRef == nil {
+		return nil, fmt.Errorf("must use TCP for gRPC listener to support HTTP API")
+	}
+	tcpAddr := *tcpAddrRef
+	tcpAddr.Port = rpcConfig.HTTPPort
+	if tcpAddr.Port == 0 {
+		tcpAddr.Port = tcpAddrRef.Port + 10
+	}
+	var listener net.Listener
+	var err error
+	if listener, err = net.ListenTCP("tcp", &tcpAddr); err != nil {
+		return nil, fmt.Errorf("failed listening for HTTP API on %v: %w", &tcpAddr, err)
+	}
+	// Close the listener if anything else in this function fails
+	success := false
+	defer func() {
+		if !success {
+			_ = listener.Close()
+		}
+	}()
+
+	// Wrap the listener in a TLS listener if there is any TLS config
+	if tlsConfigProvider != nil {
+		if tlsConfig, err := tlsConfigProvider.GetFrontendServerConfig(); err != nil {
+			return nil, fmt.Errorf("failed getting TLS config for HTTP API: %w", err)
+		} else if tlsConfig != nil {
+			listener = tls.NewListener(listener, tlsConfig)
+		}
+	}
+
+	// Create a client to frontend gRPC
+	frontendConn := rpcFactory.CreateLocalFrontendGRPCConnection()
```

Comment:
> And goes through TLS again. If you want it to stay in-process you'll need to do something different.

Yeah, I think you may be right. I am going to have the gRPC server probably listen on another listener. I don't want to use a unix socket because of platform specific concerns, I wonder if we have an in-process/memory conn/listener we use elsewhere in this repo.

> Also (at first glance) I think this bypasses authorization.. the grpc connection will look like an internal connection.

Ah I was not aware the auth checkers distinguished the two. I will need to find where that's done and make sure my local/pipe is not seen as authenticated.

> In fact if the http server terminates TLS itself (which I suppose is required) I think getting this to handle tls-based authn correctly will be a bunch of work

I'll see if I can easily do what infra does and put the client cert in an internal trusted header that I then pull back off in the mTLS authorizer. As for JWT, I figured just forwarding the headers would be good enough.

> I suppose you could stuff it in the context.

Yeah basically that.

Thanks for the feedback! I figured I needed to make these changes, I just didn't get to the tests yet.

Reply:
> Also, quick glance shows gRPC server doesn't support multiple serve calls, so I'm gonna have to have some kind of multi listener.

:(  I ran into a bunch of these limitations with the internal-frontend work and was mostly able to avoid it but I think you may have to deal with it now.

Crazy idea: grab a reference to the service and to the interceptor stack and implement interception manually? Might be easier + more efficient than messing with listeners and redoing serialization.

> I'll see if I can easily do what infra does and put the client cert in an internal trusted header (or context) that I then pull back off in the mTLS authorizer. As for JWT, I figured just forwarding the headers would be good enough.

It's not in the authorizer, it's in the authorization interceptor. And yeah, that should work for JWT. For mTLS... you can make the authorization interceptor pull out `credentials.AuthInfo` from somewhere else, but the bigger problem is that authorization plugins (which we can't control) get a context and can call `peer.FromContext` themselves (this is gross but...), and I assume you can't replace that context value since grpc will set it. Unless you do the crazy idea and bypass most of the stack and call the handler directly.

Feel free to ping me on slack, I'm (sadly) pretty familiar with this area

Reply:
> Crazy idea: grab a reference to the service and to the interceptor stack and implement interception manually? Might be easier + more efficient than messing with listeners and redoing serialization.

I'm concerned I won't get all the details right

> Feel free to ping me on slack, I'm (sadly) pretty familiar with this area

:+1: Let me get some tests going around mTLS and see what my options are here

Reply:
Ok, I updated the code to support inline gRPC invocation via direct interceptor invocation. Going to work on auth support/tests next.

Reply:
Done